tests/libvmi/vmi.go

```diff
@@ -38,6 +38,7 @@ type Option func(vmi *kvirtv1.VirtualMachineInstance)
 func New(name string, opts ...Option) *kvirtv1.VirtualMachineInstance {
 	vmi := baseVmi(name)
 
+	WithTerminationGracePeriod(0)(vmi)
```

Comment:
Why do we force it here? Are we sure it is safe to introduce it this late?

Reply:
I attempted to explain why we enforce this in the commit message:

> Users of libvmi should not to know about libvmi.DefaultTestGracePeriod. If
> they care about the grace period of their VMs, they can set it with
> WithTerminationGracePeriod(). We use WithTerminationGracePeriod(0) to
> make shutdown faster in the test. Once we migrate the termination grace
> period test to libvmi we would start using this function externally so
> we keep it exposed by the package.

I can try to rephrase: in all current usages of `libvmi`, we use `WithTerminationGracePeriod(0)` because we want a quick shutdown. A future test would want to override this value to test the `terminationGracePeriod` field.

What do you mean by "this late" ?

Reply:
I understand and agree but we should not move it to `New` as it is now a bigger change than it has to be. I personally believe that now one(reviewer/approver) should review all the `New` usage to ensure the usage of `New` stays as intended.

Also, I would expect we don't add any option to `New` as imho it should be used for 1% of the use cases. The user should specify the intended spec and we should not force any smart optimization from our side.

Reply:
By "late" I meant the fact we introduced this API and it is used. The author of tests or editor of tests could assume this method doesn't add any option and therefore it is safe to not be explicit about the `TerminationGracePeriod`

Reply:
@xpivarc , this was a default used in all factories and moved deeper to assure our tests will not use (by mistake) the system default.
I think this fits 99% of the existing and future usages, it will be an exception to see a different value set (including `nil`).

In reality, I would expect no one to care about this value, making this change accepteable.
I do understand that it is better to be explicit, and I would generally agree, but in this case the reasoning that @dankenigsberg gave regarding the exposure of default constants convinced me. Reducing the API options has a value, and this specific option does make sense to keep as an overall default.

Reply:
@EdDev I understand this does make sense for factories but this is not a factory, right? I was surprised that there is no usage of `New` outside of factories. I would still argue that `New` should be "clean" and without any options. WDYT?

Reply:
Originally I placed this assignment even deeper, in `baseVmi`, but @EdDev asked to take it a bit higher - apparently for similar reasoning. I believe that it is OK that KubeVirt has one default (30s) and the test-purposed library has another default (0s). They have a different audience and a different typical use case. I think that it would be wrong to force each and every present and future caller of `libvmi.New` to be aware of this difference, and I don't see any other way to avoid this if not by hiding it in an exported function.

Reply:
> By "late" I meant the fact we introduced this API and it is used. The author of tests or editor of tests could assume this method doesn't add any option and therefore it is safe to not be explicit about the `TerminationGracePerio

The grand majority of tests use `NewAlpine` etc, where `TerminationGracePeriod` is *already* implicitly set to zero. My PR only levels the ground, so that the very few users of vanilla `New` has the same.

Reply:
> 

I am not sure why it is more ok to have it `New` than `baseVMI`. I am going to unhold as there is no usage of `New` outside of factories but I believe it is wrong.

Reply:
I'm trying to understand what exactly you find wrong. Tests having a default that is different from what the production code has?

Reply:
Having a default in the generic constructor.

Reply:
@xpivarc so how about keeping the generic `New` pristine, and exposing a fresh factory `NewDiskless` to be used by tests? I'm not sure myself why it would be better, but name-wise it is more in line with `NewAlpine`, `NewFedora` etc. I suspect that it many other tests using `NewDiskless` would speed up boot and shutdown without any cost to the test case.

Reply:
First I find it (very) good to see such an argument (compared to ignoring quality in many other reviews at other PR/s).

To this topic, I think there is no one best solution, there will be always pros and cons.
I think the current code is good enough, even though it got added a special default. But I am also fine with Dan's proposal, although it has the cons of not being used much (therefore less a fit under `libvmi` IMO).

There is so much more to do to cleanup the code-base, lets pick up something bigger and tackle it. If one of you will prefer something else/better here, please post a PR and we can talk about it there.